tutorial/practicals/LFRic/building_code/4_psydata/Makefile

```diff
@@ -41,5 +41,21 @@ default:
 	@echo "Please use one of the transformation-specific Makefiles, e.g."
 	@echo "make -f Makefile.extract"
 
+# Target for testing the tutorials with compilation (use only one PSyData wrapper)
+compile:
+	make -f solutions/Makefile.nan_all
+
+
+# Target for CI, only do the transformation (use all transformation scripts)
+test:
+	make clean
+	make -f solutions/Makefile.extract_one time_evolution_alg_mod_psy.f90
+	make clean
+	make -f solutions/Makefile.extract_all time_evolution_alg_mod_psy.f90
+	make clean
+	make -f solutions/Makefile.nan_all time_evolution_alg_mod_psy.f90
+	make clean
+	make -f solutions/Makefile.readonly_all time_evolution_alg_mod_psy.f90
```

Comment:
I tested the `make compile` target with `gfortran` 9.3.0 and everything works except the last `readonly_all` recipe. This fails with the error:
```
gfortran -Wall -g -I ../../../../../lib/read_only/lfric -I../../../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure $(nf-config --fflags) -I../gungho_lib -c init_perturbation_kernel_mod.f90
gfortran -Wall -g -I ../../../../../lib/read_only/lfric -I../../../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure $(nf-config --fflags) -I../gungho_lib -c time_evolution_alg_mod_psy.f90
gfortran -Wall -g -I ../../../../../lib/read_only/lfric -I../../../../../src/psyclone/tests/test_files/dynamo0p3/infrastructure $(nf-config --fflags) -I../gungho_lib -c time_evolution_driver.f90
time_evolution_driver.f90:77:6:

   77 |   use time_evolution_alg_mod, only : time_evolution_alg_init, &
      |      1
Fatal Error: Cannot open module file ‘time_evolution_alg_mod.mod’ for reading at (1): No such file or directory
compilation terminated.
make: *** [Makefile.inc:108: time_evolution_driver.o] Error 1
```
(also happens when directly running line 58).

Interestingly, when I compile the previous `nan_all` recipe and don't run `make clean` in between the build succeeds. Could you please look into this issue?

Reply:
OK, this is driving me nuts - I couldn't figure out how to reproduce it. As far as I can see, ``Makefile.inc`` contains:
```
time_evolution_driver.o: time_evolution_alg_mod.o
```
So I can't really see how your make can decide to compile ``time_evolution_driver`` before ``time_evolution_alg_mod``. I tried compiling in parallel, with and without clean ... nothing.
Could you provide the following information:
1) ``make clean; rm time_evolution_alg_mod.f90``  (the ``rm`` is now fixed in my new ``clean`` target - just to be sure we are starting at the same setup).
2) ``ls`` Just so that I can see which files do exist - maybe I missed one other file or so in clean??
3) ``make compile 2>&1 | tee log-normal`` 
4) ``make -d compile 2>&1 | tee log-debug``
Then tar up and compress the two log files (esp. the latter one might be big), and see if you can attach them here, or email them to me (private address, not BOM please). Maybe that will help.

Hmm - now that I think about it - maybe also check if ``time_evolution_alg_mod.f90`` actually contains the expected module (and is not empty or so - I have seen this elsewhere, where an error caused a file to be empty, but then the makefile did not trigger a recompile).

Reply:
OK, I think I've managed to reproduce the problem - no idea what I was doing differently before, it suddenly happened. For the record, I could now reliably reproduce the problem doing:
```
make -f solutions/Makefile.readonly_all clean
make -f solutions/Makefile.readonly_all
```
which would then reproduce the error. And if I did another ``make -f solutions/Makefile.readonly_all``, then the binary would be created correctly. Hoping that this is the same problem you had, I did find a fix: I changed the PSyclone rule to be:
```
%.f90 %_psy.f90: %.x90
	psyclone $(PSYCLONE_CMD) \
	-opsy $*_psy.f90 -oalg $*.f90 $<
```
The different is that I added ``%.f90`` to the left. As an explanation, in case someone else stumbles across this (feel free to skip 🤣 ):

The ``make`` command correctly detects that it needs to create ``time_evolution_alg_mod.o`` (which is a pre-requisite for ``time_evolution_driver.o``. But without the extended rule above, it does NOT find the pattern rule to create the ``.o`` from the ``.f`` file (even though ``time_evolution_driver.f`` exists after running PSyclone). The ``make`` debug output contains:
```
    Must remake target 'time_evolution_alg_mod.o'.
    Successfully remade target file 'time_evolution_alg_mod.o'.
```
... with no commands or anything between these lines. Since ``make`` doesn't find the implicit rule (nor the explicit pattern rule in the ``makefile``), it just does nothing, assumes for reasons beyond my understanding that the file was properly created, and keeps on going. This then breaks compilation of ``time_evolution_driver.o``, since the ``time_evolution_alg_mod.o`` files does NOT exist - in spite of what ``make`` assumes.

When you then restart ``make`` (without a ``clean`` in between), the second time it finds the ``time_evolution_alg_mod.f90`` file (from the previous run), and then finds the pattern rule and compiles, and all is fine.

So I think that 'somehow' ``make`` does not look at the file system trying to find a way to create the .o file (and otherwise it can't know about the new ``time_evolution_alg_mod.f90`` file). By adding the fact that ``time_evolution_alg_mod.f90`` is also the result of the PSyclone rule, it knows about this newly created file, and then uses the implicit rule.

Alternatively, adding an explicit rule:
```
time_evolution_alg_mod.o: time_evolution_alg_mod.f
    $(F90)...
```
works as well, since (I assume) an explicit rules forces make to look at the file system again.

Catalogue this under things I never wanted to know about ``make`` :grin:

Reply:
Well, this is more than I thought I needed to know about `make` 😄 It all works now. Thank you very much for fixing and explaining it!